### ISSUE TITLE
Support something similar to export dynamic dims for torch.compile with fullgraph=True

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#144969]
### 🚀 The feature, motivation and pitch

In export, instead of using mark_dynamic on input Tensors, you can directly specify which inputs are dynamic or not

```
        dynamic_shapes = (({0: Dim("dim")}, None, None, None),)
        torch.export.export(
            Slice(),
            inp,
            dynamic_shapes=dynamic_shapes,
        )
```

We should support an analogous concept for torch.compile(fullgraph=True)

Note that we cannot easily support this when fullgraph=False, because a graph break inside the region will result in a bunch of intermediate tensors that won't have accurate dynamic/not dynamic annotations.

**Comment by zou3519** [comment]
Would be a nice programming model thing as well if both compile and export used the same Dim API
